### ISSUE TITLE
 AnimationPlayer can create duplicate keyframes at the same time position.

### ISSUE THREAD

**Issue by avencherus** [godotengine/godot#24908]
**Godot version:**

0e25c32

**OS/device including version:**

Win10 64bit

**Issue description:**

When setting a keyframe on an existing keyframe at time 0.05, instead of overwriting it, it created a new keyframe at the same time.  When dragging it away, it generates an error in the console.

Probably a rounding issue, and likely occurs at different points or from different sets of actions.

![ezgif com-video-to-gif](https://user-images.githubusercontent.com/13004169/51029546-54e83c00-159f-11e9-99b9-a6bc06635b56.gif)
![godot_master_2019-01-11_12-40-49](https://user-images.githubusercontent.com/13004169/51029634-9842aa80-159f-11e9-84ca-a7be08c2dd96.png)
![sublime_text_2019-01-11_12-42-42](https://user-images.githubusercontent.com/13004169/51029640-9a0c6e00-159f-11e9-87ca-a38fc18647bb.png)

**Steps to reproduce:**

1. Set the time step of the animation to 0.01.
2. Choose an existing keyframe, and in the track editor, manually type in 0.05 for its time.
3. Scrub the timeline time to match 0.05, and attempt to overwrite that keyframe.
4. Afterwards, drag the keyframe, and if the problem occurs you should've had 2 keyframes in the same time position.

**Comment by timoschwarzer** [comment]
Might have been another victim of `-ffast-math`...
Can you try it on a recent master build after 08fe57dcee5d82f812e346d459a509c0153f2c96?

**Comment by avencherus** [comment]
Just gave it a shot.  Unfortunately, still present in 1b5b274.  Are you getting different results?

**Comment by timoschwarzer** [comment]
No, I didn't test it. It was just my first thought... I'll take a look at this...

**Comment by avencherus** [comment]
Thanks.  X)

**Comment by avencherus** [comment]
It's hard to tell in the GIF, is it losing selection on its own?  I have noticed issues with the track information not refreshing, so probably still work needed in various places.

I'm not too clear what exactly is expected to be in p_keys, but hey this sort of looks like the problem area at least.  Thinking that this should be a float compare in there like `Math::abs(p_keys[idx - 1].time - p_time) < CMP_EPSILON)`.

I'm not too sure what's going on, but I imagine it wouldn't be good to check the current < insertion time.  If it has a floating point problem it will be something 0.1 < 0.10000001, and fall into that branch as well.

Also not sure, why it wouldn't check for > than as well.  Since the overwrite would only be interested in equal time values.

I slapped this code in and gave it a try, works at least for the issue as far as I can tell, but not sure about the purpose of the other conditions.  My knowledge here is zero, probably breaks something else.  X)

```c++
int Animation::_insert(float p_time, T &p_keys, const V &p_value) {

	int idx = p_keys.size();

	while (true) {

		// Condition for replacement.
		if(idx > 0 && Math::abs(p_keys[idx - 1].time - p_time) < CMP_EPSILON) {

			p_keys.write[idx - 1] = p_value;
			return idx - 1;

		} else { // Otherwise insert.

			p_keys.insert(idx, p_value);
			return idx;
		}

		idx--;
	}

**Comment by avencherus** [comment]
Not sure though how to editor will cope with the time value in these cases, if or when truncation to decimals might happen.

**Comment by timoschwarzer** [comment]
I'm relatively new to the Godot codebase, too, and the lack of source code comments makes it really hard to understand what's going on and how the data flow is/should be. For example, if you just drag a keyframe around, the Animation::insert method is called every time. I don't know if that's how it's supposed to be or whatever.

**Comment by akien-mga** [comment]
Fixed by #25012.
